File: Medium/C++/Maximum_Product_Subarray(152).cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) 
    {
        int n=nums.size();
        int current=1;
        int ans=INT_MIN;  
        for(int i=0;i<n;i++)
        {
            current=current*nums[i];
            ans=max(ans,current);
            
            if(nums[i]==0)
                current=1;
        }
      
        current=1;  
        for(int i=n-1;i>=0;i--)
      {
          current=current*nums[i];
          ans=max(ans,current);
           
          if(nums[i]==0)
            current=1;   
         
      }
      
        return ans;
    }
};

File: Medium/Python 3/Maximum_Product_Subarray(152).py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        totalmax, maximum,minimum = nums[0], nums[0],nums[0]
        
        for i in range(1,len(nums)):
            temp = maximum
            
            maximum = max({nums[i], nums[i]*maximum, nums[i]*minimum})
            minimum = min({nums[i], nums[i]*temp   , nums[i]*minimum})
            
            totalmax = max(totalmax, maximum)
        return totalmax